substrate/frame/staking/src/pallet/impls.rs

```diff
@@ -338,9 +338,9 @@ impl<T: Config> Pallet<T> {
 
 		let dest = Self::payee(StakingAccount::Stash(stash.clone()));
 		let maybe_imbalance = match dest {
-			RewardDestination::Controller => Self::bonded(stash)
-				.map(|controller| T::Currency::deposit_creating(&controller, amount)),
-			RewardDestination::Stash => T::Currency::deposit_into_existing(stash, amount).ok(),
+			#[allow(deprecated)]
```

Comment:
Why do we pay to stash when reward destination is controller? This means it is not migrated yet and controller and stash are different? 
It should be paid just like a RD::Account right? (we could also do a migrate of reward destination here)

Reply:
It is being paid to `stash` because the goal is to remove controller accounts completely, & in that scenario they are the same as the stash account. In a world where there are no controllers the stash acts as the controller.

Reply:
> This means it is not migrated yet and controller and stash are different?

Yes, there will be a period of time after the upgrade where the lazy migration has not processed every record yet, & if `payout_stakers` is called then the stash will be paid. I think this is a non issue as the lazy migration will be processed extremely quickly once upgrade is live.

> we could also do a migrate of reward destination here

The lazy migration call is enough, & will be the only thing to clean up later on.

Reply:
I agree its a non-issue as migration can happen pretty quickly so consider it as a nit but the logic does not seem right.

If migration has not happened, we should pay out the controller account. If migration has happened, we should never get a variant `RewardDestination::Controller` but `RewardDestination::Account(_)`. Right? (sorry for being pedantic :/ )

Reply:
The placement of controller with stash is based on the post-deprecation logic that stash and controller are the same account. I guess the migration could happen during `do_payout_stakers` but it would add a lot of additional weight to the cal given the worse case where everyone is still using a controller payee.

> If migration has happened, we should never get a variant RewardDestination::Controller but RewardDestination::Account(_). Right?

Yes, there will be no `RewardDestination::Controller` records in storage post-migration.

Reply:
I agree we can avoid migration, but if this ends up happening, paying out stash would be wrong I think. We can just keep the same logic as before (i.e. payout controller account and also add defensive check on RewardDestination::Controller variant.

Reply:
> the whole process gets dragged out longer than it needs to get rid of this stuff.

Why would it drag more? Migration is independent of payouts and a new nominator cannot create a controller anymore. 

> do it ourselves with a script or something

This is @rossbulat's plan I believe otherwise we are stuck for a long time. The migration is also free so a script would migrate them very soon just after the upgrade. So in best case we never have the defensive error.  

> Also, if they don't, they don't lose the rewards, they just go to stash which is still under their control.

They should never lose reward and they should receive the rewards in the account they declared as their reward destination  (which is the controller account in this case). Once all accounts are migrated, we will never have this arm executed so a union of `controller` and `stash` variants is a bit ugly and better solution would be to put the deprecated controller arm and have a defensive check that fails in tests. Also note, we are migrating `RewardDestination::Controller` to `RewardDestination::Account(old_controller)`, so this change should have no functional change to the logic of where rewards are paid out.

Also adding the following snipped to avoid confusion:

```
match dest {
    RewardDestination::Controller => Self::bonded(stash)
				.map(|controller| {
					defensive!("paying out controller as reward destination which is deprecated and should be migrated");
					// this should have never happened if we lazily migrated everything but if it 
					// does, just pay the controller account.
					T::Currency::deposit_creating(&controller, amount)
	}),
	// other arms ...
};
```

Reply:
Thanks for expanding @Ank4n, please see the latest (https://github.com/paritytech/polkadot-sdk/pull/2380/commits/02f5f55a883ac04212d48eedfd3dce40bedf526f), it uses the `defensive!` method.

Reply:
Looks good thanks 👍